Complete Revenant Rewrite

Xhuis opened the pull request:
:cl: Xhuis
tweak: Revenants have been renamed to umbras and have undergone some tweaks.
/ :cl:

**_To anyone who doesn't know code, this isn't actually all that daunting! It's simply just improvement of old code to make the game run better, plus the changes below.**_

I'm unhappy with the way revenants are right now, and my code for them is pretty unsatisfactory in comparison to what I know now. Although revenants will still fill the same role of just being spookier ghosts, they'll be a bit more passive - incapable, for instance, of giving diseases to people. The new revenants will be called **umbras** and will use **vitae** instead of essence.

**Total change list:**
- Revenants have been renamed to umbras. Essence has been renamed to vitae. This may be temporary.
- Umbra spawn events are now weighted higher and spawn an unoccupied umbra. Ghosts are alerted to the umbra's position and may interact with it to take control of it.
- Umbras' health is not based on vitae but has a hard cap at 100.
- Umbras have a passive vitae drain each tick, defaulting at 0.01. If the umbra runs out of vitae, they will die irrevocably. They also slowly regenerate health by doing this.
- When an umbra dies, they leave behind umbral ashes that reform after one minute. They're difficult to see and can be scattered by activating them, although they also have high research levels if you're fast enough.
- Harvesting vitae from critical targets no longer kills them. Harvesting a target in general prohibits them from being harvested until five minutes later, but they can be drained again after that.
- EMPs revitalize umbras and give them hefty amounts of vitae due to their physical nature.
- Umbras have four abilities: Toggle Nightvision, Discordant Whisper, Possess, and Thoughtsteal.
  - Toggle Nightvision is self-explanatory.
  - Discordant Whisper is identical to the original revenant's transmit.
  - Possess allows the umbra to slip into a human's body unnoticed. While in their body, umbras will slowly drain vitae from the human at a tiny rate - not enough to cause harm, but enough to induce adverse effects in the clueless human. These effects intensify over time and eventually lead to the umbra being forced out of their host.
  - Thoughtsteal paralyzes a living human for several seconds while the umbra steals their memories. After several seconds, the umbra copies the notes of the target's memories and turns invisible - the hapless victim is stunned for several seconds afterwards and can't be Thoughtstolen by the same umbra again. Umbras have an objective to steal the memories of 25% of the station's population.
- Salt piles have been added, created by salt shaker or just by splashing salt. These piles will prevent an umbra from passing and reveal them briefly if they try.

Commit:
Revenant rewrite

ChangelingRain commented:
I should note that the disease was just a fancy way to do a temporary effect that wasn't reagent-based.

MMMiracles commented:
removing all instances of revenant from the code is arguably the best first step to improving revenant

Xhuis commented:
# ayy lmao

Commit:
Removes some more revenant code but reintroduces them

Xhuis commented:
_New in this commit:_
* Removed the last of the old revenant code that I could find.
* Begins the reintroduction of revenants under the name of __umbras__, with essence being replaced by __vitae__ but still serving the same purpose.

Commit:
Actually introduces the umbras

Commit:
Adjusts a debug value

Armhulen commented:
I don't like the name Umbra. ANY other ghost name?

Xhuis commented:
I mainly changed it from "revenant" because "revenant" just means "one who returns", which, while somewhat true, also doesn't explain why they go around killing people for their life energy. I liked the term "umbra" better as it has several spooky meanings that I thought fit it better.

drynwyn commented:
I would call them Spectres, but the merest mention of Wraith: The Oblivion gives me a boner, so take that with a grain of salt.

PKPenguin321 commented:
I think people are just used to them being called revenants now but I guess really it's up to you

I'd put up a poll on the ideas forum or something

Bawhoppen commented:
probably just make a seperate pr for renaming it cause that seems like the most controversial thing

Commit:
Makes umbras playable and functional

Commit:
Grammar fix

Xhuis commented:
Cast your votes on the rename [here](https://tgstation13.org/phpBB/viewtopic.php?f=9&t=6933)!

InsaneHyena commented:
xhuis, what is it with you and overcomplicated foreign names? Why did you feel the need to replace the word "revenant" with a foreign word "umbra". I guess you could argue that you want to distance this new iteration of antag from the old one, but why would you replace the word "essence" with "vitae"?

Xhuis commented:
I simply thought that the name __umbra_ fit better. As for _vitae_ that's Latin for _life_ and I thought it sounded better than _essence_, which I though sounded kind of lame anyway.

ChangelingRain commented on interface/stylesheet.dm:
```diff
@@ -91,13 +93,6 @@ h1.alert, h2.alert		{color: #000000;}
-.revenbignotice			{color: #1d2953;	font-weight: bold; font-size: 3;}
-.revenminor				{color: #823abb}
-.revenwarning			{color: #760fbb;	font-style: italic;}
-.revendanger			{color: #760fbb;	font-weight: bold; font-size: 3;}
```
There were a few other things using these spanclasses.

Xhuis replied:
God damn it.

X-TheDark commented:
Why the passive vitae drain? I thought you wanted to make them more passive, not force them to act (which is what vitae drain does, in my opinion).

coiax commented:
> umbra
> foreign word
> late 16th century (denoting a phantom or ghost): from Latin, literally ‘shade’.

Umbra is in the english language, used as a literary word for shadow or darkness or as a technical photography term.

> revenant
> early 19th century: French, literally ‘coming back’, present participle (used as a noun) of revenir .

The moral of the story is they're both "foreign" words, just like the other 58% of words that come from Latin and French influence.

![image](https://cloud.githubusercontent.com/assets/609465/16005808/28051f0c-3161-11e6-8401-42799fa8b1a6.png)

coiax commented:
Also :+1: to renaming them and using these names, anything to break ties with Griefghost.

Xhuis commented:
@X-TheDark I felt like the original revenants could wander around and kind of do whatever the hell they wanted, and I didn't want to have the same issue with umbras. They're not going to be as direct on the round, but they'll still need to stock up on vitae every so often.

There may be abilities or passive unlocks that counter the drain later on, but vitae will generally only be gainable through harvesting or perhaps abilities.

There are no planned abilities right now other than a re-introduction of Transmit.

RemieRichards commented on code/game/gamemodes/miniantags/umbra/phantasmal_ashes.dm:
```diff
@@ -0,0 +1,44 @@
+
+/obj/item/phantasmal_ashes/New()
+	..()
+	spawn(UMBRA_ASHES_REFORM_TIME)
```
user ```addtimer()```

Xhuis replied:
I have no idea what that is.

X-TheDark replied:
In this case, the addtimer() syntax would be (I think):
`addtimer(src, "reform", UMBRA_ASHES_REFORM_TIME)`

The addtimer takes care of the case when src is deleted. For general description, you can take a look at https://github.com/tgstation/tgstation/blob/master/code/controllers/subsystem/timer.dm#L53

Instead of spawn(), it defines your code that you want to run delayed as an "event" (think of it as a random event that occurs in-game or the running of process(). Here, you define the thing to call (usually src), the proc (from src) to call, the time when it needs to be called (same number as you'd use in spawn()).

RemieRichards commented on code/game/gamemodes/miniantags/umbra/umbra.dm:
```diff
@@ -0,0 +1,247 @@
+
+/mob/living/simple_animal/umbra/Life()
+	..()
+	adjust_vitae(-UMBRA_VITAE_DRAIN_RATE, TRUE, "passive drain")
```
```addtimer()```

RemieRichards commented on code/game/gamemodes/miniantags/umbra/umbra.dm:
```diff
@@ -0,0 +1,247 @@
+	visible_message("<span class='warning'>[src] winks out of existence, releasing its hold on [L]...</span>")
+	L.visible_message("<span class='warning'>...who falls unceremoniously back to the ground.</span>")
+	animate(L, pixel_y = pixel_y - 5, time = 10)
+	spawn(rand(UMBRA_MAX_HARVEST_COOLDOWN - 600, UMBRA_MAX_HARVEST_COOLDOWN)) //Between 4-5 minutes by default
```
```addtimer()``` pls

RemieRichards commented on code/game/gamemodes/miniantags/umbra/umbra.dm:
```diff
@@ -0,0 +1,247 @@
+	src << "<span class='warning'>You've become visible!</span>"
+	alpha = 255
+	invisibility = FALSE
+	spawn(time)
```
```addtimer()```

RemieRichards commented on code/game/gamemodes/miniantags/umbra/umbra.dm:
```diff
@@ -0,0 +1,247 @@
+		return
+	src << "<span class='warning'>You can't move!</span>"
+	notransform = TRUE
+	spawn(time)
```
yadda yadda

Commit:
Some more work

Xhuis commented:
_New in this commit:_
* Added an `isumbra()` helper proc.
* Added relevant umbra spans to `stylesheet.dm.`
* Added three abilities to umbras.
 * Toggle Nightvision is self-explanatory.
 * Discordant Whisper functions like Transmit once did.
 * Manifest causes spooky effects and generates a little bit of vitae for everyone nearby.
    * An invulnerable effect is created for a few seconds, plus loud sounds and a red screen flash.
    * Effects are determined by the umbra's Manifest theme, set using a verb in the IC tab.
    * Effects include a vaguely man-shaped hole in reality speaking to everyone nearby or a spooky blood-covered clown making distorted honks.
* Added a 1% chance for umbras to be named grief ghosts.
* Added and edited some variables for umbras to help them work as intended.

JJRcop commented on code/game/gamemodes/miniantags/umbra/umbra.dm:
```diff
@@ -0,0 +1,247 @@
+		vitae_yield += rand(10, 15)
+	if(L.stat == UNCONSCIOUS)
+		vitae_information += "blazing with vitality, "
+		vitae_yield += rand(20, 25) //Significant bonus if the target is in critical condition instead of dead
```
This comment is kinda weird, because you can easily be unconscious without being in crit

Xhuis replied:
Good point, let me change that

Commit:
Fixes harvesting working on sleeping mobs

Cheridan commented:
>I don't like the name Umbra. ANY other ghost name?

I still say that you should call it a Grudge. And then we can all say that you're Grudgecoding.

X-TheDark commented on code/game/gamemodes/miniantags/umbra/umbra_mob.dm:
```diff
@@ -0,0 +1,275 @@
+	visible_message("<span class='warning'>[src] winks out of existence, releasing its hold on [L]...</span>")
+	L.visible_message("<span class='warning'>...who falls unceremoniously back to the ground.</span>")
+	animate(L, pixel_y = pixel_y - 5, time = 10)
+	spawn(rand(UMBRA_MAX_HARVEST_COOLDOWN - 600, UMBRA_MAX_HARVEST_COOLDOWN)) //Between 4-5 minutes by default
```
Here, you would use the following syntax:
```
var/time_to_call = rand(UMBRA_MAX_HARVEST_COOLDOWN - 600, UMBRA_MAX_HARVEST_COOLDOWN)
addtimer(src, "remove_from_list", time_to_call, FALSE, L)
```
After which, I think you'd do something like:
```
/mob/living/simple_animal/umbra/proc/remove_from_list(mob/living/L)
    if(!L)
        return
    src << "<span class='umbra'>You think that [L]'s soul should be strong enough to harvest again.</span>"
    recently_drained -= L
```

Commit:
Removes Manifest

Xhuis commented:
_New in this commit:_
* Replaces a lot of `spawn()`s with `addtimer()`s.
* Removed Manifest, as I was unhappy with it.

coiax commented:
> some spawns left in place

Okay, tommorow I will make a PR and burn all the spawns to the ground

Xhuis commented:
I meant for stuff like the `Stun()` and `Reveal()` procs. I'm not sure why I'd need to make procs for unstunning and hiding again.

Commit:
It compiles

X-TheDark commented:
Would it be possible to have a quick shortcut to cast Discordant Whisper (such as Ctrl+Clicking a mob). Because sometimes, trying to transmit to a target in a crowded area when the target decides to not stand still can be an issue.

Xhuis commented:
@X-TheDark Should be easy enough.

Commit:
Changes fluff to be more sci-fi

Xhuis commented:
_New in this commit:_
* Heavily modified the fluff thanks to ideas from @Shadowlight213.
 * Umbras themselves are semi-corporeal entities capable of conscious thought and kept together by an electromagnetic field that constantly decays.
 * _Vitae_ is a substance produced by living things in times of distress and used to energize and sustain them. Umbras drain this from corpses or those in critical condition to sustain themselves. Once a being is stable again, this vitae is absorbed back into their body for later use.

If there's any holes in that caused by my ignorance of how electromagnetic fields work, let me know.

lordpidey commented:
@Xhuis , sintouched and soulsellers are separate things.  Also consider using the mind.soulowner variable.

Commit:
Self-healing and grammar tweak

lzimann commented:
is it really going to be called umbra now(i didn't read anything in the pr tbh)

Xhuis commented:
Polls are showing that most people prefer revenant so that's probably what I'll go with in the end.

Commit:
Adds possession

Commit:
Conflicts 1

Commit:
Conflicts 2

Commit:
Conflicts 3

Commit:
Forces umbras out of dead possession targets

Xhuis commented:
_New in this commit:_
* Fixed conflicts.
* Added salt piles. Created by using a salt shaker on the ground or by splashing salt onto the ground, they will prevent the passage of umbras and briefly reveal them if they try.
* Umbras are now immune to things like the singularity, Nar-Sie, and Ratvar.
* Added Possession. Umbras can now undetectably enter a human's body to slowly leech vitae at a small rate. The umbra cannot move while in this form, but can still use Discordant Whisper. While possessing a target, the umbra will gain vitae at a rate that it normally drained and the target will experience adverse effects. These effects will worsen until, after a certain amount of time (five minutes by default), the umbra is forced out. An umbra can unpossess its target at any time by using Possession again and are also stunned, revealed, and forced out if their target has salt in their system.

Bawhoppen commented:
since you're cucking around were you going to make them be damaged by ocular wardens like I suggested earlier

Kromgar commented:
Ocular wardens have truesight!

Xhuis commented:
Ocular wardens can already see and attack umbras - I have no intention of changing that.

Bawhoppen commented:
they already could???? ok

Commit:
Changes some fluff and animation

Commit:
Adds objectives and such

Xhuis commented:
_New in this commit:_
* Renamed phantasmal ashes to umbral ashes.
* Added Thoughtsteal, a new ability with a 45s cooldown. The umbra appears, stunning and immobilizing the chosen human. After five uninterrupted seconds, the umbra tears off a portion of the target's brain and consumes it, copying the target's notes for itself. This can't be done more than once on the same target.
* Umbras now have the objective to steal the thoughts of a number of targets equal to 25% of the round population.
* Added an event to spawn umbras, set to occur semi-frequently up to ten minutes into the round.

This should be merge-ready!

optimumtact commented on code/game/gamemodes/miniantags/umbra/umbra.dm:
```diff
@@ -0,0 +1,369 @@
+//Creation, destruction, life, and death
+/mob/living/simple_animal/umbra/New()
+	..()
+	if(prob(1))
```
I do not think this is a good idea

Xhuis replied:
:(

optimumtact replied:
I won't block this PR however

Xhuis replied:
<3

optimumtact commented on code/game/gamemodes/miniantags/umbra/umbra.dm:
```diff
@@ -0,0 +1,369 @@
+/mob/living/simple_animal/umbra/death()
+	if(breaking_apart)
+		return
+	..(1)
```
1?

Xhuis replied:
Gibbed - so it doesn't leave behind a body

optimumtact commented on code/modules/mob/mob_helpers.dm:
```diff
@@ -445,8 +445,9 @@ It's fairly easy to fix if dealing with single letters but not so much with comp
-	log_admin("[key_name(usr)] has offered control of ([key_name(M)]) to ghosts.")
-	message_admins("[key_name_admin(usr)] has offered control of ([key_name_admin(M)]) to ghosts")
+	if(usr)
+		log_admin("[key_name(usr)] has offered control of ([key_name(M)]) to ghosts.")
```
was usr ever unset in your testing?

Xhuis replied:
Yes, during early development. I threw this sanity check in to make sure it wouldn't happen again.

Kromgar commented:
What if instead of copying it steals the notes?

Suddenly a traitor no longer remembers they are an antag.

Would be hard to enforce but damn would it be interesting.

That or reshuffle their pda code and reset the pda upon having your notes stolen

KorPhaeron commented:
>passive drain

I liked just chilling and chatting with people in the bar though.

Spawning on low pop and then dying because there arent any bodies doesnt sound fun

KorPhaeron commented:
Can you please put all the changes in one place (the op)

KorPhaeron commented:
https://www.tgstation13.org/phpBB/viewtopic.php?f=9&t=6546

Majority of people like revenant as is anyway, not sure why it needw to be rewritten

Xhuis commented:
I mainly wanted to rewrite the shoddy code that made up old revenants and decided to take the opportunity to make them better fit the original vision I had for them. I'll post changes in the OP>

Bawhoppen commented:
what if you rewrite revenant code, leave it the same, then add a new mob that does all these new things

Xhuis commented:
I suppose that would work, but I feel like that'd make two very similar mobs that accomplish the same goal of being half-ghosts that disrupt the station.

Bawhoppen commented:
enemy halfghosts

Bawhoppen commented:
just maybe

ChangelingRain commented on code/game/gamemodes/miniantags/revenant/revenant_blight.dm:
```diff
@@ -1,68 +0,0 @@
-		if(5)
-			if(!finalstage)
-				finalstage = 1
-				affected_mob << "<span class='revenbignotice'>You feel like [pick("nothing's worth it anymore", "nobody ever needed your help", "nothing you did mattered", "everything you tried to do was worthless")].</span>"
```
I'd be very sad if all this text went away, I like making the clown have a bad time.

Xhuis replied:
I might add it in as a sort of side effect of having your thoughts stolen.

optimumtact commented:
Death to the salt ghost

Bawhoppen commented:
oranges this is what I'm talking about with you merging things way too prematurely

optimumtact commented:
It's not prematurely the code was ready, go find some other shitty side event antag to annoy players with.

Bawhoppen commented:
it was prematurely???? there was still discussion about the name aswell as other balance concerns you cuckwad

Xhuis commented:
That's why we have post-merge patches, Bawhop.

optimumtact commented:
you ranting because your favourite annoying tactic got taken away is not discussion. Go away bawhoppen, go away and stay away.

Bawhoppen commented:
my favorite annoying tactic? what is that exactly? seriously oranges, stop being a bitch and assuming, you should realise that merging stuff too quickly is a bad idea, I don't give a fuck about the contents of the PR

AdamElTablawy commented:
this seems fairly unfinished

Xhuis commented:
I consider it to be done. If you have any feedback I'd be glad to hear it.

jughu commented:
You can see umbras with the medscanner and beat them whike invinsibke
